src/lib.rs

```diff
@@ -135,15 +136,13 @@ pub trait Socket {
     /// connections on it.
     ///
     /// Returns the bound-to endpoint, with the port resolved (if applicable).
-    async fn bind(&mut self, endpoint: impl TryIntoEndpoint + 'async_trait)
-        -> ZmqResult<&Endpoint>;
+    async fn bind(&mut self, endpoint: impl TryIntoEndpoint + 'async_trait) -> ZmqResult<Endpoint>;
 
     // TODO: Although it would reduce how convenient the function is, taking an
     // `&Endpoint` would be better for performance here
     async fn connect(&mut self, endpoint: impl TryIntoEndpoint + 'async_trait) -> ZmqResult<()>;
 
-    /// Retrieves the list of currently bound endpoints
-    fn binds(&self) -> &[Endpoint];
+    fn binds(&self) -> &HashMap<Endpoint, oneshot::Sender<bool>>;
```

Comment:
My concern here is that we are exposing in the public api the implementation details. Ideally this would be an iterator of some sort. It also forces us to have one representation of the hashmap for all the different socket types - how do we know that a different socket type might not need a value of a different type, unique to that socket type? Customizing the type of the hashmap or the iterator is probably impossible without generic associated types, so this may become a real issue later. 

Its definitely not a blocker for this MR as its more important to fix the resource leak - I can try to address it in a followup PR

Reply:
I agree that this is not the best option and we may need to rework it later. For now I've just implemented a quick-fix that solves 1 particular issue without big changes in other parts.